content/blog/2022-04-07-compiling-black-with-mypyc-part-2.md

```diff
@@ -0,0 +1,537 @@
+
+#### Tightening up existing type annotations
+
+The stricter your type annotations are in your codebase, the more invariants mypyc will be
+able to infer. It'll then use this information and write type-specific code that is
+faster. This code won't work if it gets an object of a different type, but that's why we
+use and enforce type annotations, so we (and mypyc) can safely assume it's going to be the
+right type.
+
+This involves reading through the code and control flow, checking whether certain states
+are impossible. Blib2to3 is mostly a legacy codebase, so we when added type annotations to
+it, it was done to unblock other work. The goal was to make blib2to3 type check, and not
+write the best typed code ever. So, naturally there were a few permissive (parameter) type
+annotations I could make stricter:
+
+```diff
+diff --git a/src/blib2to3/pgen2/parse.py b/src/blib2to3/pgen2/parse.py
+index 47c8f02..6b03188 100644
+--- a/src/blib2to3/pgen2/parse.py
++++ b/src/blib2to3/pgen2/parse.py
+@@ -138,7 +138,7 @@ class Parser(object):
+         self.rootnode: Optional[NL] = None
+         self.used_names: Set[str] = set()
+
+-    def addtoken(self, type: int, value: Optional[Text], context: Context) -> bool:
++    def addtoken(self, type: int, value: Text, context: Context) -> bool:
+         """Add a token; return True iff this is the end of the program."""
+         # Map from token to label
+         ilabel = self.classify(type, value, context)
+@@ -185,11 +185,10 @@ class Parser(object):
+                     # No success finding a transition
+                     raise ParseError("bad input", type, value, context)
+
+-    def classify(self, type: int, value: Optional[Text], context: Context) -> int:
++    def classify(self, type: int, value: Text, context: Context) -> int:
+         """Turn a token into a label.  (Internal)"""
+         if type == token.NAME:
+             # Keep a listing of all used names
+-            assert value is not None
+             self.used_names.add(value)
+             # Check for reserved words
+             ilabel = self.grammar.keywords.get(value)
+@@ -201,12 +200,10 @@ class Parser(object):
+         return ilabel
+
+     def shift(
+-        self, type: int, value: Optional[Text], newstate: int, context: Context
++        self, type: int, value: Text, newstate: int, context: Context
+     ) -> None:
+         """Shift a token.  (Internal)"""
+         dfa, state, node = self.stack[-1]
+-        assert value is not None
+-        assert context is not None
+         rawnode: RawNode = (type, value, context, None)
+         newnode = self.convert(self.grammar, rawnode)
+         if newnode is not None:
+```
+
+This also has the neat side-effect of allowing me to remove some asserts.
+
+Tightening up type annotations involving [`typing.Any`][typing-any] can be particularly
+worthwhile as it forces mypyc to fallback to generic C code that can handle any kind of
```

Comment:
```suggestion
worthwhile as `Any` forces mypyc to fallback to generic C code that can handle any kind of
```
Otherwise "it" sounds like it refers back to "Tightening"